internal/k8s/statusupdater.go

```diff
@@ -0,0 +1,186 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// StatusUpdate contains an all the information needed to change an object's status to perform a specific update.
+// Send down a channel to the goroutine that actually writes the changes back.
+type StatusUpdate struct {
+	FullName FullName
+	GVR      schema.GroupVersionResource
+	Mutator  StatusUpdateMutator
+}
+
+// StatusUpdateMutator is an interface to hold mutator functions for status updates.
+type StatusUpdateMutator interface {
+	Mutate(obj interface{}) interface{}
+}
+
+// StatusUpdateHandler holds the details required to actually write an Update back to the referenced object.
+type StatusUpdateHandler struct {
+	Log           logrus.FieldLogger
+	Clients       *Clients
+	UpdateChannel chan StatusUpdate
+	LeaderElected chan struct{}
+	IsLeader      bool
+	Converter     *UnstructuredConverter
+}
+
+// Start runs the goroutine to perform status writes.
+// Until the Contour is elected leader, will drop updates on the floor.
+func (sw *StatusUpdateHandler) Start(stop <-chan struct{}) error {
+
+	for {
+		select {
+		case <-stop:
+			return nil
+		case <-sw.LeaderElected:
+			sw.Log.Info("elected leader")
+			sw.IsLeader = true
+			// disable this case
+			sw.LeaderElected = nil
+		case upd := <-sw.UpdateChannel:
+			if !sw.IsLeader {
+				sw.Log.WithField("fullname", upd.FullName).Debug("not leader, not applying update")
+				continue
+			}
+
+			sw.Log.WithField("fullname", upd.FullName).Debug("Received a status update to write.")
+			uObj, err := sw.Clients.DynamicClient().Resource(upd.GVR).Namespace(upd.FullName.Namespace).Get(context.TODO(), upd.FullName.Name, metav1.GetOptions{})
+			if err != nil {
+				sw.Log.WithError(err).WithField("fullname", upd.FullName).WithField("groupversionresource", upd.GVR).Error("unable to retrieve object for updating")
+				continue
+			}
+
+			obj, err := sw.Converter.FromUnstructured(uObj)
+			if err != nil {
+				sw.Log.WithError(err).WithField("fullname", upd.FullName).Error("unable to convert from unstructured")
+				continue
+			}
+
+			newObj := upd.Mutator.Mutate(obj)
+
+			if assert.StatusEqual(obj, newObj) {
+				sw.Log.WithField("fullname", upd.FullName).Debug("Update was a no-op")
+				continue
+			}
+
+			usNewObj, err := sw.Converter.ToUnstructured(newObj)
+			if err != nil {
+				sw.Log.WithError(err).WithField("fullname", upd.FullName).Error("unable to convert update to unstructured")
+				continue
+			}
+
+			_, err = sw.Clients.DynamicClient().Resource(upd.GVR).Namespace(upd.FullName.Namespace).UpdateStatus(context.TODO(), usNewObj, metav1.UpdateOptions{})
+			if err != nil {
+				sw.Log.WithError(err).WithField("fullname", upd.FullName).Error("unable to update status")
+				continue
+			}
+
+		}
+
+	}
+
+}
+
+// StatusUpdater describes an interface to send status updates somewhere.
+type StatusUpdater interface {
+	SetStatus(name, namespace string, gvr schema.GroupVersionResource, mutator StatusUpdateMutator)
```

Comment:
Suggest that the natural operation for a status updater is `Update`.

Reply:
Did you consider using a `FullName` for the first argument?

Reply:
Renaming to Update and using a `FullName` are both great ideas, thanks.

Reply:
Hmm, `Ingress` objects don't have an accessor that gives you the `FullName`, I don't think that changing to `FullName` here adds much except the need to make a new `FullName{}` struct whenever we call `Update`.